lib/compress/zstdmt_compress.c

```diff
@@ -1232,6 +1243,112 @@ static size_t ZSTDMT_flushProduced(ZSTDMT_CCtx* mtctx, ZSTD_outBuffer* output, u
     return 0;   /* internal buffers fully flushed */
 }
 
+/**
+ * Returns the range of data used by the earliest job that is not yet complete.
+ * If the data of the first job is broken up into two segments, we cover both
+ * sections.
+ */
+static range_t ZSTDMT_getInputDataInUse(ZSTDMT_CCtx* mtctx)
+{
+    unsigned const firstJobID = mtctx->doneJobID;
+    unsigned const lastJobID = mtctx->nextJobID;
+    unsigned jobID;
+
+    for (jobID = firstJobID; jobID < lastJobID; ++jobID) {
+        unsigned const wJobID = jobID & mtctx->jobIDMask;
+        size_t consumed;
+
+        ZSTD_PTHREAD_MUTEX_LOCK(&mtctx->jobs[wJobID].job_mutex);
+        consumed = mtctx->jobs[wJobID].consumed;
+        ZSTD_pthread_mutex_unlock(&mtctx->jobs[wJobID].job_mutex);
+
+        if (consumed < mtctx->jobs[wJobID].src.size) {
+            range_t range = mtctx->jobs[wJobID].prefix;
+            if (range.size == 0) {
+                /* Empty prefix */
+                range = mtctx->jobs[wJobID].src;
+            } else if (mtctx->jobs[wJobID].src.start < range.start) {
+                /* Two segments */
+                assert(mtctx->jobs[wJobID].src.start == mtctx->inBuffSpace.start);
+                range.start = mtctx->inBuffSpace.start;
+                range.size = mtctx->inBuffSpace.capacity;
```

Comment:
In this case, the "range of data used" covers the entire `inBuffSpace`, which means everything is in use, which means no space is available to load next job.

I'm wondering if that's a necessary restriction.

It seems enough to get the beginning of earliest job not yet completed,
which is basically `mtctx->jobs[wJobID].prefix.start`.

The assumption here is that this is a round buffer.
In which case `nextAvailableStart` immediately follows a busy segment.
What matters then is to check if there is enough "free" space in front of `nextAvailableStart`.

If `nextAvailableStart <= earliestJobStart`, the assumption is that the whole space between these 2 positions is available for loading new data, and we just have to check that's large enough to load a job.

If `nextAvailableStart > earliestJobStart`, the assumption is that everything from `nextAvailableStart` to the end of `inBuffSpace` is available to load new data, though we still have to check if that's large enough to load a job.
If not, then `nextAvailableStart = inBuffSpace.start`, and it becomes equivalent to previous case.

So, in case "earliest job not completed" spans the end and beginning of `inBuffSpace`, it's still possible to reserve some space to load some new job.

Feel free to comment.

Reply:
Since we don't support 2-segment jobs yet, I'm just going to assert that it isn't the case.